senders/mattermost/mattermost.go

```diff
@@ -0,0 +1,173 @@
+package mattermost
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/moira-alert/moira"
+	"github.com/moira-alert/moira/senders"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+type Sender struct {
+	frontURI string
+	location *time.Location
+	client   Client // client is interface to be possible to stub Mattermost model.Client4 dependency.
+}
+
+func NewSender() *Sender {
+	return &Sender{}
+}
+
+// Init configures Sender.
+func (sender *Sender) Init(senderSettings map[string]string, _ moira.Logger, location *time.Location, _ string) error {
+	url := senderSettings["url"]
+	if url == "" {
+		return fmt.Errorf("can not read Mattermost url from config")
+	}
+	sender.client = model.NewAPIv4Client(url)
+
+	token := senderSettings["api_token"]
+	if token == "" {
+		return fmt.Errorf("can not read Mattermost api_token from config")
+	}
+	sender.client.SetToken(token)
+
+	sender.frontURI = senderSettings["front_uri"]
+	sender.location = location
+
+	// TODO make GET request to get bot ID to send direct messages.
+
+	return nil
+}
+
+// SendEvents implements moira.Sender interface.
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, _ [][]byte, throttled bool) error {
+	message := sender.buildMessage(events, trigger, throttled)
+	err := sender.sendMessage(message, contact.Value, trigger.ID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (sender *Sender) buildMessage(events moira.NotificationEvents, trigger moira.TriggerData, throttled bool) string {
+	const messageMaxCharacters = 4_000
+
+	var message strings.Builder
+
+	title := sender.buildTitle(events, trigger)
+	titleLen := len([]rune(title))
+
+	desc := sender.buildDescription(trigger)
+	descLen := len([]rune(desc))
+
+	eventsString := sender.buildEventsString(events, -1, throttled)
+	eventsStringLen := len([]rune(eventsString))
+
+	charsLeftAfterTitle := messageMaxCharacters - titleLen
+
+	descNewLen, eventsNewLen := senders.CalculateMessagePartsLength(charsLeftAfterTitle, descLen, eventsStringLen)
+
+	if descLen != descNewLen {
+		desc = desc[:descNewLen] + "...\n"
+	}
+	if eventsNewLen != eventsStringLen {
+		eventsString = sender.buildEventsString(events, eventsNewLen, throttled)
+	}
+
+	message.WriteString(title)
+	message.WriteString(desc)
+	message.WriteString(eventsString)
+	return message.String()
+}
+
+func (sender *Sender) buildDescription(trigger moira.TriggerData) string {
+	desc := trigger.Desc
+	if trigger.Desc != "" {
+		desc += "\n"
+	}
+	return desc
+}
+
+func (sender *Sender) buildTitle(events moira.NotificationEvents, trigger moira.TriggerData) string {
+	title := fmt.Sprintf("*%s*", events.GetSubjectState())
+	triggerURI := trigger.GetTriggerURI(sender.frontURI)
+	if triggerURI != "" {
+		title += fmt.Sprintf(" <%s|%s>", triggerURI, trigger.Name)
+	} else if trigger.Name != "" {
+		title += " " + trigger.Name
```

Comment:
strings concatenate will be faster with strings builder

Reply:
Message title is pretty short, so strings.Builder is premature optimization